feat(tools): add sentry to client for error handling

TristanToye opened the pull request:
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #39751

Got this to log to a free sentry account with a gitpod running. Followed similar pattern to server. I am unclear on if there is a better place to put this integration for the project setup. Also wondering if there is a known client side error I could trigger here more easily than creating a fake one for testing steps?

TristanToye committed:
Add Sentry to client for error handling

raisedadead commented:
@ojeytonwilliams is there anything preventing us from getting this in?

ojeytonwilliams commented:
No, I don't have any reservations. It just needs reviewing and the lock files updating (I'll handle that since we've left this sitting for so long).

raisedadead commented:
Perfect. Thanks a lot.

nhcarrigan commented on client/src/html.tsx:
```diff
@@ -1,4 +1,18 @@
 import React from 'react';
+import { sentry } from '../../config/secrets';
+
+if (sentry.dns === 'dsn_from_sentry_dashboard') {
+  console.log('Sentry reporting disabled unless DSN is provided!');
+} else {
+  Sentry.init({
+    dsn: sentry.dns,
+    integrations: [new Integrations.BrowserTracing()],
+    tracesSampleRate: 1.0
+  });
+  console.log('Sentry initialized');
```
Do we want to be using the same DSN for both the client and API?

ojeytonwilliams replied:
Good catch. No, that would be very confusing.

I've set up a new project and I'll make that key available to the client.

ojeytonwilliams committed:
feat: switch to Gatsby plugin

The react sdk wasn't working so, rather than digging into it, I used
the plugin.

ShaunSHamilton commented on client/src/utils/report-error.ts:
```diff
@@ -1,7 +1,13 @@
-  message = 'Unhandled error'
-): void {
-  return console.error(`Client: ${message}`, e);
+  message = sentry.dns === 'dsn_from_sentry_dashboard'
```
Can we fix this in `/config/secrets.js`? That is, make it `sentry.dsn`?

ojeytonwilliams replied:
My typo has survived two years so far, but sure.

nhcarrigan replied:
I propose `sentry.dsn$`

ojeytonwilliams replied:
```js
$(function() {
 Sentry.init({
    dsn: sentry.dsn$,
    integrations: [new Integrations.BrowserTracing()],
    tracesSampleRate: 1.0
  });
});
```

ojeytonwilliams committed:
fix: use env.json in report-error

ojeytonwilliams committed:
refactor: dns -> dsn

ojeytonwilliams committed:
fix: update import

moT01 reviewed (approved):
LGTM @TristanToye 🎉 I know this has been sitting here for a while, but thanks for working on this and congratulations on your first contribution to freeCodeCamp 🎉 

I suppose we need to add some keys to the builds - anyone want to volunteer?

ojeytonwilliams commented:
> I suppose we need to add some keys to the builds - anyone want to volunteer?

Sure, I'll do that now.
